### PR TITLE
Vary the pruning behavior and target based on type of druid

### PR DESCRIPTION
**Pull request by mjgiarlo**
## Why was this change made? 🤔

This commit patches a bug discovered in dor-services-app where the assumption that a druid should be pruned starting at its parent path leads to the pruning of Access druids deleting unrelated objects (because Access druids lack the leaf ID node that other druids have).

## How was this change tested? 🤨

CI

### CODE REVIEW

File: spec/druid_tools_spec.rb [hunk 1]
```diff
@@ -115,25 +117,41 @@
...
 
     druid2.rmdir
     expect(File.exist?(tree1)).to be true
     expect(File.exist?(tree2)).to be false
+    expect(File.exist?(tree3)).to be false
 
     druid1.rmdir
     expect(File.exist?(tree1)).to be false
     expect(File.exist?(tree2)).to be false
+    expect(File.exist?(tree3)).to be false
```
Comment (mjgiarlo) [inline spec/druid_tools_spec.rb:154]: These methods (rmdir, mkdir, prune! below) are deprecated, yes, but these test changes establish that acting on a single druid does not affect other druids.

File: spec/druid_tools_spec.rb [hunk 2]
```diff
@@ -322,6 +341,32 @@
...
+        expect(File).to exist(dr1.pruning_base.parent)
+      end
+
+      it 'stops at ancestor directories that have children' do
+        # 'cd/456/ef' should still exist because of dr1
+        shared_ancestor = dr1.pruning_base.parent
+        expect(shared_ancestor.to_s).to match(%r{cd/456/ef$})
+        expect(File).to exist(shared_ancestor)
+      end
+    end
```
Comment (mjgiarlo) [inline spec/druid_tools_spec.rb:368]: This is the same test change made in https://github.com/sul-dlss/dor-services-app/pull/3790

File: spec/druid_tools_spec.rb (hunk 1 above)
Comment (ndushay) [reply spec/druid_tools_spec.rb:154]: maybe put this comment in the spec code?

File: spec/druid_tools_spec.rb [hunk 3]
```diff
@@ -6,6 +6,8 @@
   let(:tree1) { File.join(fixture_dir, 'cd/456/ef/7890/cd456ef7890') }
   let(:strictly_valid_druid_str) { 'druid:cd456gh1234' }
   let(:tree2) { File.join(fixture_dir, 'cd/456/gh/1234/cd456gh1234') }
+  let(:access_druid_str) { 'druid:cd456ef0987' }
```
Comment (ndushay) [inline spec/druid_tools_spec.rb:9]: could the last 4 digits be really different, like 9999?

**Review by ndushay — APPROVED** [review]

File: spec/druid_tools_spec.rb (hunk 1 above)
Comment (mjgiarlo) [reply spec/druid_tools_spec.rb:154]: @ndushay I'll add some words to the `it` string above.

File: spec/druid_tools_spec.rb (hunk 3 above)
Comment (mjgiarlo) [reply spec/druid_tools_spec.rb]: yup! done.

**Review by jmartin-sul — COMMENTED** [review]
LGTM, but code climate is complaining about lack of coverage for the added/modified methods, which i don't really understand, given the tests.  the fact that the tests look for side effects on the file system makes it seem like there wouldn't be inadvertent method mocking/interception getting in the way of things.

Screen Shot 2022-04-01 at 11 50 25 AM

File: spec/druid_tools_spec.rb [hunk 4]
```diff
@@ -312,6 +330,7 @@
     let(:workspace) { Dir.mktmpdir }
     let(:dr1) { described_class.new(druid_str, workspace) }
     let(:dr2) { described_class.new(strictly_valid_druid_str, workspace) }
+    let(:dr3) { described_class.new(access_druid_str, workspace) }
```
Comment (jmartin-sul) [inline spec/druid_tools_spec.rb:333]: actually, should this be a `DruidTools::AccessDruid`?

**Review by jmartin-sul — CHANGES_REQUESTED** [review]
ok, i can see how maybe `DruidTools::AccessDruid` isn't getting covered, but i still don't quite understand the coverage complaint about `DruidTools::Druid`.

File: spec/druid_tools_spec.rb (hunk 4 above)
Comment (mjgiarlo) [reply spec/druid_tools_spec.rb:333]: Yep. Making this and the other changes above now. Thank you!

Comment (mjgiarlo) [reply spec/druid_tools_spec.rb]: @jmartin-sul This is ready for another look.

**Review by ndushay — APPROVED** [review]
loving it.

**Review by jmartin-sul — APPROVED** [review]
thanks for the touchups!  i pulled the branch and inserted some logging statements and re-ran the tests, and i think i'm now convinced that codeclimate is confused?

```
druid-tools % git diff                                                             druid-pruning-base
diff --git a/lib/druid_tools/access_druid.rb b/lib/druid_tools/access_druid.rb
index 419b7e0..8fd01e2 100644
--- a/lib/druid_tools/access_druid.rb
+++ b/lib/druid_tools/access_druid.rb
@@ -17,6 +17,7 @@ module DruidTools
     end
 
     def pruning_base
+      puts "running (#{self.class}\##{__method__})"
       pathname
     end
   end
diff --git a/lib/druid_tools/druid.rb b/lib/druid_tools/druid.rb
index 6a248e5..a6a0472 100644
--- a/lib/druid_tools/druid.rb
+++ b/lib/druid_tools/druid.rb
@@ -165,10 +165,12 @@ module DruidTools
     end
 
     def pruning_base
+      puts "running (#{self.class}\##{__method__})"
       pathname.parent
     end
 
     def prune!
+      puts "running (#{self.class}\##{__method__})"
       pruning_base.rmtree if pruning_base.exist? && pruning_base != base_pathname
       prune_ancestors pruning_base.parent
     end
druid-tools %
druid-tools % be rspec                                                             druid-pruning-base
[Coveralls] Using SimpleCov's default settings.
.running (DruidTools::AccessDruid#pruning_base)
.........running (DruidTools::Druid#pruning_base)
.......................running (DruidTools::Druid#prune!)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::Druid#prune!)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::AccessDruid#prune!)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
.running (DruidTools::AccessDruid#prune!)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::AccessDruid#prune!)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::AccessDruid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::Druid#prune!)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::Druid#prune!)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.running (DruidTools::Druid#prune!)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
running (DruidTools::Druid#pruning_base)
.

Finished in 0.08124 seconds (files took 1.29 seconds to load)
41 examples, 0 failures

[Coveralls] Outside the CI environment, not sending data.
druid-tools %
```

**Pull request merged**